### PR TITLE
[Bugfix] #7749 order with status cancelled/brought_by_himself/done unable to edit

### PR DESCRIPTION
**Pull request by yuliiakaras**
[#7749](https://github.com/ita-social-projects/GreenCity/issues/7749)

### CODE REVIEW

**Review by kryzanivska-nastya — APPROVED** [review]

File: src/app/shared/address-input/address-input.component.ts [hunk 1]
```diff
@@ -43,12 +43,13 @@ import { AddressService } from '../services/address/address.service';
...
   ]
 })
-export class AddressInputComponent implements OnInit, AfterViewInit, OnDestroy, ControlValueAccessor, Validator {
+export class AddressInputComponent implements OnInit, AfterViewInit, OnDestroy, OnChanges, ControlValueAccessor, Validator {
   @Input() edit: boolean;
   @Input() address: Address;
   @Input() addFromProfile: boolean;
   @Input() isShowCommentInput = true;
   @Input() isFromAdminPage: boolean;
+  @Input() uneditableStatus: boolean;
```
Comment (Lokankara) [inline src/app/shared/address-input/address-input.component.ts]: The prefix `is` or `has` is a common convention for `boolean` variables

File: src/app/ubs/ubs-admin/components/ubs-admin-export-details/ubs-admin-export-details.component.ts [hunk 2]
```diff
@@ -81,9 +85,9 @@ export class UbsAdminExportDetailsComponent implements OnInit, OnDestroy, OnChan
   get isFormRequired(): boolean {
     const isNotOpen = !this.pageOpen;
     const isNotValid = !this.exportDetailsDto.valid;
-    const isNotCancelOrDone = !this.isOrderStatusCancelOrDone;
+    const isUneditable = !this.uneditableStatus;
 
-    return isNotOpen && isNotValid && isNotCancelOrDone;
+    return isNotOpen && isNotValid && isUneditable;
```
Comment (Lokankara) [inline src/app/ubs/ubs-admin/components/ubs-admin-export-details/ubs-admin-export-details.component.ts]: Is it possible to simplify the method and avoid using intermediate constants?

File: src/app/ubs/ubs-admin/components/ubs-admin-export-details/ubs-admin-export-details.component.html [hunk 3]
```diff
@@ -2,17 +2,17 @@
   <div class="form" [formGroup]="exportDetailsDto">
     <div class="empty_card" (click)="openDetails()">
       <img
-        *ngIf="!pageOpen && (exportDetailsDto.valid || isOrderStatusCancelOrDone)"
```
Comment (Lokankara) [inline src/app/ubs/ubs-admin/components/ubs-admin-export-details/ubs-admin-export-details.component.html:5]: simply call the getArrowImageSrc() method for the src attribute of the  tag. 
 This removes the need for multiple *ngIf statements:
 `arrow`

File: src/app/ubs/ubs-admin/components/ubs-admin-table/table-cell-date/table-cell-date.component.html [hunk 4]
```diff
@@ -1,12 +1,12 @@
...
   <span *ngIf="date !== '-'">{{ date | date: 'dd/MM/y' }}</span>
-  <span *ngIf="!isEditable && !isBlocked" class="pointer" (click)="edit()">&#9998;</span>
+  <span *ngIf="!isEditable && !isBlocked && !uneditableStatus" class="pointer" (click)="edit()" (keydown)="edit($event)">&#9998;</span>
   <mat-spinner *ngIf="isBlocked" strokeWidth="3" diameter="20"></mat-spinner>
 </div>
-<mat-form-field *ngIf="isEditable" color="accent" appearance="fill">
+
+<mat-form-field *ngIf="isEditable && !uneditableStatus" color="accent" appearance="fill">
   <input matInput [matDatepicker]="picker1" [min]="current" [value]="date" (dateChange)="changeData($event)" id="date-input" />
   <mat-datepicker-toggle matSuffix [for]="picker1" color="accent"></mat-datepicker-toggle>
```
Comment (Lokankara) [inline src/app/ubs/ubs-admin/components/ubs-admin-table/table-cell-date/table-cell-date.component.html:11]: Attribute color is not allowed here

File: src/app/ubs/ubs-admin/components/ubs-admin-table/ubs-admin-table.component.ts [hunk 5]
```diff
@@ -98,7 +98,7 @@ export class UbsAdminTableComponent implements OnInit, AfterViewChecked, OnDestr
   isStoreEmpty: boolean;
   isPostData = false;
   dataForPopUp = [];
```
Comment (Lokankara) [inline src/app/ubs/ubs-admin/components/ubs-admin-table/ubs-admin-table.component.ts:99]: Unused fields: previousIndex, mouseEvents, minColumnWidth
Unused methods: checkIfFilteredBy, clearFilters

**Review by Lokankara — APPROVED** [review]

**Review by kovalsofiia1 — APPROVED** [review]

File: src/app/ubs/ubs-admin/components/ubs-admin-responsible-persons/ubs-admin-responsible-persons.component.html [hunk 6]
```diff
@@ -1,17 +1,17 @@
 <div class="container" [formGroup]="responsiblePersonsForm">
   <div class="empty_card" (click)="openDetails()">
     <img
-      *ngIf="!pageOpen && (responsiblePersonsForm.valid || isOrderStatusCancelOrDone)"
+      *ngIf="!pageOpen && (responsiblePersonsForm.valid || uneditableStatus)"
```
Comment (bzhn) [inline src/app/ubs/ubs-admin/components/ubs-admin-responsible-persons/ubs-admin-responsible-persons.component.html]: Complex conditions may be refactored for better readability and maintainability.

Also, you can check out the following conversation with ChatGPT to learn more how you could enhance the code:
https://chatgpt.com/share/674cdc0a-1b8c-800a-b940-d92948f324f1

File: src/app/shared/address-input/address-input.component.ts (hunk 1 above)
Comment (bzhn) [inline src/app/shared/address-input/address-input.component.ts]: Agree with @Lokankara. I would use `isUneditableStatus` or `canEditStatus` names. The current name is a bit confusing, personally, for me.

**Review by bzhn — APPROVED** [review]

File: src/app/ubs/ubs-admin/components/ubs-admin-responsible-persons/ubs-admin-responsible-persons.component.html (hunk 6 above)
Comment (bzhn) [reply src/app/ubs/ubs-admin/components/ubs-admin-responsible-persons/ubs-admin-responsible-persons.component.html]: Unfortunately, the link is not valid anymore. But here is the content.

The given Angular template code toggles between displaying different images based on conditions. Here are some observations and suggestions to validate or enhance the implementation:

### Observations
1. **Conditional Logic**: The logic uses `*ngIf` to determine which `` tag should render based on the component's state (`pageOpen`, `responsiblePersonsForm.valid`, and `uneditableStatus`).
2. **Code Readability**: While the code achieves the intended behavior, repeated conditions make it harder to read and maintain.
3. **Asset Paths**: Ensure the paths (`assets/img/ubs-admin-orders/...`) are correct and the assets exist in the specified locations.
4. **Alt Text Accessibility**: The `alt` attributes lack descriptive text for screen readers.

### Suggestions for Improvement
#### 1. **Use a Single Conditional Logic Block**
You can replace the three `` tags with one, binding the `src` attribute dynamically. This reduces redundancy and makes the code more readable.

```html
<img
  [src]="getArrowImageSrc()"
  [alt]="getArrowImageAlt()"
/>
```

In the component's TypeScript file:
```typescript
getArrowImageSrc(): string {
  if (!this.pageOpen && (this.responsiblePersonsForm.valid || this.uneditableStatus)) {
    return 'assets/img/ubs-admin-orders/arrow_up.svg';
  } else if (this.pageOpen && (this.responsiblePersonsForm.valid || this.uneditableStatus)) {
    return 'assets/img/ubs-admin-orders/arrow_down.svg';
  } else if (this.pageOpen && !this.responsiblePersonsForm.valid && !this.uneditableStatus) {
    return 'assets/img/ubs-admin-orders/red_arrow_down.svg';
  }
  return ''; // Fallback if no condition is met
}

getArrowImageAlt(): string {
  if (!this.pageOpen && (this.responsiblePersonsForm.valid || this.uneditableStatus)) {
    return 'Arrow pointing up';
  } else if (this.pageOpen && (this.responsiblePersonsForm.valid || this.uneditableStatus)) {
    return 'Arrow pointing down';
  } else if (this.pageOpen && !this.responsiblePersonsForm.valid && !this.uneditableStatus) {
    return 'Red arrow pointing down';
  }
  return 'No arrow'; // Fallback
}
```

#### 2. **Accessibility Improvements**
- The `alt` attributes should provide meaningful descriptions for users with visual impairments. The suggested `getArrowImageAlt` function addresses this.
- Ensure these descriptions are contextually relevant to the images' function or meaning.

#### 3. **Refactor Complex Conditions**
If the conditions are reused elsewhere or are too complex, consider abstracting them into utility functions:
```typescript
isArrowUp(): boolean {
  return !this.pageOpen && (this.responsiblePersonsForm.valid || this.uneditableStatus);
}

isArrowDown(): boolean {
  return this.pageOpen && (this.responsiblePersonsForm.valid || this.uneditableStatus);
}

isRedArrowDown(): boolean {
  return this.pageOpen && !this.responsiblePersonsForm.valid && !this.uneditableStatus;
}
```

Then, use these in the `getArrowImageSrc` and `getArrowImageAlt` methods.

#### 4. **Error Handling**
- Add a fallback or default `src` value to handle scenarios where none of the conditions are met.
- Log or handle unexpected states, especially in larger, dynamic applications.

By implementing these suggestions, the code will be more maintainable, readable, and accessible.

**Review by hnativlyubomyr — APPROVED** [review]

**Pull request merged**